syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.product";
option java_outer_classname = "ProductProto";
option objc_class_prefix = "HLW";

package product;

message ProductReply {
    int32 id = 1;
    string name = 2;
}

message ProductsReply {
    repeated ProductReply products = 1;
}

message CreateRequest {
    string supplierName = 1;
}

message UpdateRequest {
    int32 id = 1;
    string supplierName = 2;
}

message FindByIdRequest {
    int32 id = 1;
}

message PaginationRequest {
    int32 page = 1;
    int32 skip = 2;
}

message Empty {}

service Product {
    rpc FindAllProduct (Empty) returns (ProductsReply) {}
    rpc FindOneProduct (FindByIdRequest) returns (ProductReply) {}
    rpc CreateProduct (CreateRequest) returns (ProductReply) {}
    rpc UpdateProduct (UpdateRequest) returns (ProductReply) {}
    rpc RemoveProduct (FindByIdRequest) returns (ProductReply) {}
    rpc QueryProduct (stream PaginationRequest) returns (stream ProductReply) {}
}